### PR TITLE
StormCast: Training improvements and code refactoring

### PR DESCRIPTION
**Pull request by albertocarpentieri**
## Description

This MR introduces several improvements to the StormCast training pipeline, focusing on configurability, reproducibility, and code maintainability.

### New features

* Added build_optimizer function supporting multiple optimizer types: adam, adamw, stableadamw
* Optimizer parameters (betas, weight_decay, eps, fused) are now configurable via YAML
* Added support for torch schedulers
* Helper functions init_scheduler and step_scheduler for scheduler management

### Performance

Added perf configuration section (similar to CorrDiff example) with:
  * fp_optimizations: Support for fp32, amp-fp16, amp-bf16
  * use_apex_gn: Apex GroupNorm with channels_last memory format

### Training loop

* Refactored monolithic training_loop function into a Trainer class for better readability and maintainability
* Reproducible validation with fixed seeding
* Configurable number of validation steps
* Validation loss saved in checkpoint metadata
* Gradient clipping support
* TensorBoard logging support

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies
torch-optimi

### CODE REVIEW

File: physicsnemo/models/diffusion_unets/song_unet.py [hunk 1]
```diff
@@ -873,6 +880,7 @@ def __init__(
         N_grid_channels: int = 4,
         checkpoint_level: int = 0,
         additive_pos_embed: bool = False,
+        bottleneck_attention: bool = True,
```
Comment (pzharrington) [inline physicsnemo/models/diffusion_unets/song_unet.py:883]: Document this arg in the docstring like in `SongUNet`

File: physicsnemo/models/diffusion_unets/song_unet.py [hunk 2]
```diff
@@ -1497,6 +1506,7 @@ def __init__(
         prob_channels: List[int] = [],
         checkpoint_level: int = 0,
         additive_pos_embed: bool = False,
+        bottleneck_attention: bool = True,
```
Comment (pzharrington) [inline physicsnemo/models/diffusion_unets/song_unet.py:1509]: Document this arg in the docstring like in `SongUNet`

File: physicsnemo/models/diffusion_unets/song_unet.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# SPDX-FileCopyrightText: Copyright (c) 2023 - 2026 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2025 NVIDIA CORPORATION & AFFILIATES.
```
Comment (pzharrington) [inline physicsnemo/models/diffusion_unets/song_unet.py]: Update the year header

File: examples/weather/stormcast/utils/io.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
-# limitations under the License.
+# limitations under the License..
```
Comment (pzharrington) [inline examples/weather/stormcast/utils/io.py]: Remove the extra period here

File: physicsnemo/models/diffusion_unets/song_unet.py (hunk 2 above)
Comment (albertocarpentieri) [reply physicsnemo/models/diffusion_unets/song_unet.py:1509]: Here we say: Most parameters are similar to the parent
    :class:`~physicsnemo.models.diffusion_unets.SongUNetPosEmbd`, at the
    exception of the ones listed below.

**Review by pzharrington — APPROVED** [review]
LGTM, most review was done in gitlab. Will address the examples docstring formatting after Jussi's updates are merged

File: physicsnemo/models/diffusion_unets/song_unet.py [hunk 5]
```diff
@@ -178,6 +178,9 @@ class SongUNet(Module):
         temporal information about the diffusion process. In that sense it is a
         simpler version of the positional embedding used in
         :class:`~physicsnemo.models.diffusion_unets.SongUNetPosEmbd`.
+    bottleneck_attention : bool, optional, default=True
+        If ``True``, applies self-attention at the bottleneck (innermost decoder block).
+        Set to ``False`` to disable bottleneck attention for faster inference.
```
Comment (CharlelieLrt) [inline physicsnemo/models/diffusion_unets/song_unet.py:183]: @pzharrington Do you know what's the difference between this attention on that already implemented?

Comment (pzharrington) [reply physicsnemo/models/diffusion_unets/song_unet.py:183]: It is the one that's implemented, this is just making it configurable. See below, previously it was just hardcoded:

Image
